xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -173,58 +199,40 @@ static String extractUdsTarget(ConfigSource configSource) {
 
     @Override
     public void onNext(DiscoveryResponse discoveryResponse) {
-      try {
-        processDiscoveryResponse(discoveryResponse);
-      } catch (Exception e) {
-        logger.log(Level.SEVERE, "processDiscoveryResponse", e);
-      }
-    }
 
-    private void processDiscoveryResponse(DiscoveryResponse response)
-        throws InvalidProtocolBufferException {
-      List<Any> resources = response.getResourcesList();
-      ArrayList<String> resourceNames = new ArrayList<>();
-      for (Any any : resources) {
-        final String typeUrl = any.getTypeUrl();
-        checkState(SECRET_TYPE_URL.equals(typeUrl), "wrong value for typeUrl %s", typeUrl);
-        Secret secret = Secret.parseFrom(any.getValue());
-        resourceNames.add(secret.getName());
-        processSecret(secret);
-      }
-      lastResponse = response;
-      // send discovery request as ACK
-      sendDiscoveryRequestOnStream(resourceNames);
-    }
+      processDiscoveryResponse(discoveryResponse);
 
-    private void processSecret(Secret secret) {
-      synchronized (watcherMap) {
-        final HashSet<SecretWatcher> secretWatchers = watcherMap.get(secret.getName());
-        if (secretWatchers != null) {
-          for (SecretWatcher secretWatcher : secretWatchers) {
-            secretWatcher.onSecretChanged(secret);
-          }
-        }
-      }
     }
 
     @Override
     public void onError(Throwable t) {
-      logger.log(Level.SEVERE, "ResponseObserver.onError", t);
+      sendErrorToWatchers(t);
     }
 
+
+
     @Override
     public void onCompleted() {
       // TODO(sanjaypujare): add retry logic once client implementation is final
     }
   }
 
-  private void sendDiscoveryRequestOnStream(String... names) {
-    sendDiscoveryRequestOnStream(Arrays.asList(names));
+  private void processDiscoveryResponse(DiscoveryResponse response) {
+    try {
+      processSecretsFromDiscoveryResponse(response);
+    } catch (Exception e) {
+      sendErrorToWatchers(e);
+      // send Nack
+      sendNack(Code.INTERNAL_VALUE, e.getMessage());
```

Comment:
For `com.google.rpc.Status` (which is used here) its Builder takes only int value and *NOT* `Code` enum. Also instead of asking callers to build `com.google.rpc.Status` I thought this is better.